aten/src/ATen/native/quantized/cpu/fbgemm_utils.h

```diff
@@ -55,4 +62,606 @@ inline void convert_int8_uint8(
   }
 }
 
+// Calculate the column offsets.
+// Note this includes the sum of the columns as well as the scalar term
+// B_zero_point * K, whereas the row_offsets created by
+// PackAWithQuantRowOffset is only the sum of the A rows.
+inline void calc_col_offsets_transpose(
+    int K,
+    int N,
+    const int8_t* Bint8,
+    int32_t* B_zero_point,
+    int32_t* col_offsets,
+    c10::QScheme qtype) {
+  for (size_t i = 0; i < N; ++i) {
+    int32_t sum = 0;
+    for (size_t j = 0; j < K; ++j) {
+      sum += Bint8[i * K + j];
+    }
+    if (qtype == c10::kPerTensorAffine) {
+      col_offsets[i] = sum - B_zero_point[0] * K;
+    } else {
+      col_offsets[i] = sum - B_zero_point[i] * K;
+    }
+  }
+}
+
+#if AT_MKLDNN_ENABLED()
+static at::Tensor mkldnn_linear_prepack(
+    at::Tensor& weight,
+    c10::optional<at::Tensor> bias) {
+  TORCH_CHECK(
+      weight.dim() == 2,
+      "The weight tensor for quantized::linear_prepack should"
+      " be 2-dimensional.");
+  auto N = weight.size(0);
+  // TODO: contiguous is called for further JIT optimizations.
+  auto weight_contig = weight.contiguous();
+  const auto qtype = weight.qscheme();
+
+  c10::optional<at::Tensor> bias_contig;
+  if (bias.has_value()) {
+    at::Tensor bias_vec = bias.value();
+    TORCH_CHECK(bias_vec.dim() == 1, "bias should be a vector (1D Tensor)");
+    TORCH_CHECK(
+        bias_vec.size(0) == N,
+        "bias should have N elements: " + std::to_string(N));
+    bias_contig = bias->contiguous();
+  }
+
+  std::vector<int32_t> col_offsets(N);
+  
+  std::vector<int32_t> weight_zero_points_int32(1, 0);
+  if (qtype == c10::kPerTensorAffine) {
+    weight_zero_points_int32[0] = weight.q_zero_point();
+  } else if (qtype == c10::kPerChannelAffine) {
+    weight_zero_points_int32.resize(N, 0);
+    for (int i = 0; i < N; ++i) {
+      weight_zero_points_int32[i] =
+          weight.q_per_channel_zero_points()[i].item<int32_t>();
```

Comment:
this is pretty slow actually as indexing creates a new tensor (didn't notice in trunk). Best it to get `.quantizer()` and cast it to the corresponding type (something like `q_per_channel_scales_quant`'s implementation does)

Reply:
done